### DISCUSSION TITLE
Centralized Contract Pipeline

### DISCUSSION THREAD

**Discussion by l-monninger** [movementlabsxyz/movement#552] [Ideas]
# Summary
One of the top priorities for our centralized launch is ensuring our contract pipeline is reasonably stable and secure. The purpose of this discussion is to work towards a specification of the contract pipeline design.

## Initial Proposal
![pipeline](https://github.com/user-attachments/assets/2b3a0581-cc4c-4cf8-a84c-3b549599c4a3)

1. Contract upgrade workflow MUST be triggered from a commit on main or a tag. 
2. Contract upgrade workflow MUST not continue if any tests fail.
3. Contract upgrade workflow MUST not attempt deploy new implementation contracts if the implementation has not changed.
4. Contract upgrade workflow MUST trigger AWS KMS managed signers for a multisig that can execute Timelocked upgrade of the proxy contract. 

**Outstanding Questions w/ Approach**:
- How do we want to handle devnet and testnet? Should this all be based on patterned tags? 
- Do we need to complete #400 for consistency of tagging?
- What break-glass measures should we put in place?

**Comment by l-monninger** [comment]
<img width="416" alt="Screenshot 2024-09-10 at 3 20 17 PM" src="https://github.com/user-attachments/assets/6945cece-5fc4-48e4-85f2-a01b22d609e9">

What do we want to use to implement this in particular? Custom script? Something from OZ?

**Reply by Primata** [reply]
Are we automating upgrades? Once a contract change happens, proposal is pushed and eventually upgrade is executed? This seems a little bit too much, I think that we can manually run proposals and executions if that's ever happening. Worried that an accidental push to main could trigger an upgrade and open up an attack vector.

**Reply by l-monninger** [reply]
> Once a contract change happens, proposal is pushed and eventually upgrade is executed?

Essentially. I'm not suggesting it has to be exactly the above in terms of how that pattern is actually implemented though. 

> Are we automating upgrades? 

Yes, the objective is to automate, so that we don't end up relying on operators, directly exposing keys, etc. At some point, the tags or commits on main are planned to be used as governance for the canonical chain. 

>  Worried that an accidental push to main could trigger an upgrade and open up an attack vector.

This would also be subject to intervention at the time-lock in the same way a manual upgrade would.

**Reply by Primata** [reply]
Do we want a 48hrs timelock for all contracts then?

**Reply by Primata** [reply]
I think the cleanest solution is to keep track of changes in main for contracts and trigger upgrades if there are changes. A foundry script can cleanly perform these upgrades, I'm just wondering how the enclave could pick up the movement github repo and run the upgrade.

Since I know you guys like versioning, my suggestion would be:
First version of the contract deployed would be manual
Subsequent deployments are picked up by contract and version:
StakingV2.sol:StakingV2 aka. {contract-name}V{version}.sol: {contract-name}V{version}

This means that the branch coming in has to be named upgrade-{contract-name}-V{version} and contract names do not have to stay the same but have to follow and standard for an upgrade to happen.

**Reply by l-monninger** [reply]
>  I'm just wondering how the enclave could pick up the movement github repo and run the upgrade.

This wouldn't be that hard. But, we won't be doing enclaves yet.

**Reply by l-monninger** [reply]
> Do we want a 48hrs timelock for all contracts then?

I think the best way to answer this is by comparison. What is working at other well-run orgs should work at ours.

**Reply by l-monninger** [reply]
> This means that the branch coming in has to be named upgrade-{contract-name}-V{version} and contract names do not have to stay the same but have to follow and standard for an upgrade to happen.

I would prefer to do this on a tag. But, yes, tagging with the unit upgraded is probably a good idea for a mono.

**Reply by Primata** [reply]
I know a few protocols that do 48hrs.

Versioning is kept with historical onchain data. I do prefer calling whatever was Legacy and the latest being just the contract name. Older versions are held onchain and it also prevents people from using older versions.

**Reply by Primata** [reply]
I think that other orgs are not really automating to this degree because there are too many variables in place and this ends up being counter productive. Just changing the fact that you'd like to make a call during an upgradeAndCall breaks the whole thing.

**Reply by l-monninger** [reply]
> I know a few protocols that do 48hrs.

Let's provide a list. 

In theory, this kind of thing is also actually good for an MIP, because it covers governance standards we adopt for ourselves as maintainers of the canonical chain.

**Reply by l-monninger** [reply]
> I think that other orgs are not really automating to this degree because there are too many variables in place and this ends up being counter productive. Just changing the fact that you'd like to make a call during an upgradeAndCall breaks the whole thing.

Deployments should also be tested in the CI. Which orgs are you referring to though? If you're talking about things like DEXs, the bar is a bit different.

**Comment by l-monninger** [comment]
@0xPrimata, let's talk about a possible tagging strategy, because a lot of the concerns with automation I have and I'm guessing you have can be attenuated via the usage of tags and releases.

Have you seen the Aptos release strategy?: https://github.com/aptos-labs/aptos-core/blob/main/RELEASE.md

**Reply by l-monninger** [reply]
@0xPrimata 

> This means that the branch coming in has to be named upgrade-{contract-name}-V{version} and contract names do not have to stay the same but have to follow and standard for an upgrade to happen.

I would prefer to do this on a tag. But, yes, tagging with the unit upgraded is probably a good idea for a mono.
